ambry-router/src/test/java/com.github.ambry.router/RouterTestHelpers.java

```diff
@@ -140,11 +140,22 @@ static void setServerErrorCodes(ServerErrorCode[] serverErrorCodesInOrder, Parti
     if (serverErrorCodesInOrder.length > replicas.size()) {
       throw new IllegalArgumentException("More server error codes provided than replicas in partition");
     }
+    // Do serverErrorCodes assignment based on Operation Tracker logic so that server error can return in the order of
+    // serverErrorCodesInOrder.
+    int numOfNodesEachDatacenter = 3;
+    ServerErrorCode[] serverErrorCodesInOperationTrackerOrder = new ServerErrorCode[serverErrorCodesInOrder.length];
+    for (int i = 0; i < numOfNodesEachDatacenter; i++) {
+      serverErrorCodesInOperationTrackerOrder[i] = serverErrorCodesInOrder[numOfNodesEachDatacenter - 1 - i];
+    }
+    for (int i = numOfNodesEachDatacenter; i < serverErrorCodesInOrder.length; i++) {
+      serverErrorCodesInOperationTrackerOrder[i] =
+          serverErrorCodesInOrder[serverErrorCodesInOrder.length + numOfNodesEachDatacenter - 1 - i];
+    }
     int i = 0;
     for (ReplicaId replica : partition.getReplicaIds()) {
       DataNodeId node = replica.getDataNodeId();
       MockServer mockServer = serverLayout.getMockServer(node.getHostname(), node.getPort());
-      mockServer.setServerErrorForAllRequests(serverErrorCodesInOrder[i++]);
+      mockServer.setServerErrorForAllRequests(serverErrorCodesInOperationTrackerOrder[i++]);
```

Comment:
Maybe you can change this loop in the format of `for(int i = 0; i < partition.getReplicaIds().size(); ++i)`. Thus the order is guaranteed in a safer way. (also guarantee the order of mockServer/node)

Reply:
use `serverErrorCodesInOperationTrackerOrder.get(i++)` if you adopt my suggestion in previous comment.

Reply:
fixed.